zproject/backends.py

```diff
@@ -449,79 +430,84 @@ def ldap_to_django_username(self, username):
 class ZulipLDAPAuthBackend(ZulipLDAPAuthBackendBase):
     REALM_IS_NONE_ERROR = 1
 
-    def authenticate(self, username: str, password: str, realm_subdomain: Optional[Text]=None,
+    def authenticate(self, username: str, password: str, realm: Optional[Realm]=None,
                      return_data: Optional[Dict[str, Any]]=None) -> Optional[UserProfile]:
+        if realm is None:
+            return None
+        self._realm = realm
+        if not ldap_auth_enabled(realm):
+            return None
+
         try:
-            self._realm = get_realm(realm_subdomain)
             username = self.django_to_ldap_username(username)
-            user_profile = ZulipLDAPAuthBackendBase.authenticate(self,
-                                                                 username=username,
-                                                                 password=password)
-            if user_profile is None:
-                return None
-            if not user_matches_subdomain(realm_subdomain, user_profile):
-                return None
-            return user_profile
-        except Realm.DoesNotExist:
-            return None  # nocoverage # TODO: this may no longer be possible
+            return ZulipLDAPAuthBackendBase.authenticate(self,
+                                                         username=username,
+                                                         password=password)
         except ZulipLDAPException:
             return None  # nocoverage # TODO: this may no longer be possible
 
     def get_or_create_user(self, username, ldap_user):
         # type: (str, _LDAPUser) -> Tuple[UserProfile, bool]
-        try:
-            if settings.LDAP_EMAIL_ATTR is not None:
-                # Get email from ldap attributes.
-                if settings.LDAP_EMAIL_ATTR not in ldap_user.attrs:
-                    raise ZulipLDAPException("LDAP user doesn't have the needed %s attribute" % (
-                        settings.LDAP_EMAIL_ATTR,))
-
-                username = ldap_user.attrs[settings.LDAP_EMAIL_ATTR][0]
-
-            user_profile = get_user_profile_by_email(username)
-            if not user_profile.is_active or user_profile.realm.deactivated:
-                raise ZulipLDAPException("Realm has been deactivated")
-            if not ldap_auth_enabled(user_profile.realm):
-                raise ZulipLDAPException("LDAP Authentication is not enabled")
+
+        if settings.LDAP_EMAIL_ATTR is not None:
+            # Get email from ldap attributes.
+            if settings.LDAP_EMAIL_ATTR not in ldap_user.attrs:
+                raise ZulipLDAPException("LDAP user doesn't have the needed %s attribute" % (
+                    settings.LDAP_EMAIL_ATTR,))
+
+            username = ldap_user.attrs[settings.LDAP_EMAIL_ATTR][0]
+
+        return_data = {}  # type: Dict[str, Any]
+        user_profile = common_get_active_user(username, self._realm, return_data)
+        if user_profile is not None:
```

Comment:
I intentionally didn't do that, because if you do, it makes that previous commit harder to read.

Reply:
(Added a commit message explanation to this effect)

Reply:
Hmm -- that's what `git diff -b` is for, no? Or in gitk it's even easier, with the "Ignore space change" toggle at the top of the diff. So I think I'd still prefer this fixup squashed in, though it's not a real strong preference.

Reply:
Merged, with a tweak to the "some unnecessary indentation" commit message too.

Reply:
`git diff -b` is nice for that, but because many people try to review things on e.g. the GitHub UI where that isn't available, I'm still in the habit of trying to split code reformatting from logic changes in adjacent commits wherever convenient.